tools/config/config.go

```diff
@@ -0,0 +1,321 @@
+package config
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+	"k8s.io/client-go/util/connrotation"
+	"k8s.io/klog/v2"
+	"net"
+	"net/http"
+	"reflect"
+	"sigs.k8s.io/cluster-inventory-api/apis/v1alpha1"
+	"sigs.k8s.io/cluster-inventory-api/tools/plugins"
+	"sigs.k8s.io/cluster-inventory-api/tools/plugins/spiffe"
+	"sync"
+	"time"
+)
+
+const (
+	APIServerEndpointPropertyKey = "cluster-entrypoints.k8s.io"
+	CABunblePropertyKey          = "cabundle.k8s.io"
+	PluginPropertyKey            = "plugin.multicluster.k8s.io"
+)
+
+var pluginMap = map[string]plugins.PluginInterface{}
+
+func init() {
+	spiffePlugin, err := spiffe.NewPlugin()
+	if err == nil {
+		pluginMap[spiffePlugin.Name()] = spiffePlugin
+	} else {
+		utilruntime.HandleError(err)
+	}
+}
+
+// BuildConfigFromClusterProfile is to build the rest.Config to init the client.
+func BuildConfigFromClusterProfile(cluster *v1alpha1.ClusterProfile) (*rest.Config, error) {
+	// get the plugin by clusterprofile
+	plugin, err := getPluginFromClusterProfile(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterEndpoint, err := getEndpointFromClusterProfile(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	config := &rest.Config{
+		Host: clusterEndpoint,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: getCABundleFromClusterProfile(cluster),
+		},
+	}
+
+	a := newAuthenticator(plugin, cluster)
+	transportConfig, err := config.TransportConfig()
+	if err := a.UpdateTransportConfig(transportConfig); err != nil {
+		return nil, err
+	}
+
```

Comment:
I'd merge all this as "apply plugin for credentials" on the config (so that it's all put together)